### PR TITLE
223-446 Restructure IO for Graph-ODE

### LINKED ISSUE #223: Add option to avoid output of single run results
If a user runs a lot of ensemble runs are done, the output of the single run results which blow up the needed disk space might not be wanted and only output of percentiles is seeked. We should implement an option to avoid outputting data for every single run.

Another interesting option would be to provide the user an estimate of the used disk space.

### LINKED ISSUE #364: Create ABM simulation with percentiles output
As extension of the current, ABM example we want to provide a simulation scenario where multiple runs are possible and where then simulation percentiles are output.

This issue is blocked by:
- #463 
- #464
- #465

### LINKED ISSUE #446: Rearrange results output in Graph-ODE simulation
A lot of result IO is currently done in the simulation files itself. This has to be ported to either general locations (memilio/...) or to specific ones (model/...)

### PR DESCRIPTION
**Pull request by mknaranja**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [x] Tests for new functionality has been added
* [x] A local test was succesful
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)
* [ ] If new third party software is used, did you pay attention to its license? Please remember to add it to the wiki after successful merging.
* [ ] If new mathematical methods or epidemiological terms are used, has the glossary been updated ? Did you provide further documentation ?
 is present or referenced. Please provide your references.
* [ ] The following questions are addressed in the documentation*:  Developers (what did you do?, how can it be maintained?), For users (how to use your work?), For admins (how to install and configure your work?)
* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [x] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [x] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [x] There are appropriate **unit tests** and they pass.
* [x] The git history is clean and linearized for the merge request.
* [x] Coverage report for new code is acceptable. 

Closes #446
Closes #223

### CODE REVIEW

File: cpp/memilio/io/result_io.h [hunk 1]
```diff
@@ -86,11 +91,121 @@ class SimulationResult
...
+    auto ensemble_result_sum = sum_nodes(ensemble_results);
+    auto num_groups          = (int)(size_t)ensemble_params[0][0].parameters.get_num_groups();
+    if(save_single_runs)
+    {
+        for (size_t i = 0; i < ensemble_result_sum.size(); ++i) {
+        BOOST_OUTCOME_TRY(save_result(ensemble_result_sum[i], {0}, num_groups,
+                                           (result_dir / ("results_run" + std::to_string(i) + "_sum.h5")).string()));
+        BOOST_OUTCOME_TRY(save_result(ensemble_results[i], county_ids, num_groups,
+                                           (result_dir / ("results_run" + std::to_string(i) + ".h5")).string()));
+        }
```
Comment (mknaranja) [inline cpp/memilio/io/result_io.h:126]: @HenrZu Let's have a look on this on Monday. Isn't this actually already saved in the run function?

File: cpp/memilio/io/result_io.h [hunk 2]
```diff
@@ -86,11 +93,141 @@ class SimulationResult
...
+ * Creates a new subdirectory for each run according to run_idx.
+ * @param result Simulation results per node of the graph.
+ * @param params Parameters used for the simulation run.
+ * @param ids Identifiers for each node of the graph. 
+ * @param result_dir Top level directory for all results of the parameter study.
+ * @param run_idx Index of the run; used in directory name.
+ * @return Any io errors that occur during writing of the files.
+ */
+template <class Model>
+IOResult<void> save_result_with_params(const std::vector<TimeSeries<double>>& result, const std::vector<Model>& params,
```
Comment (mknaranja) [inline cpp/memilio/io/result_io.h:112]: needs test

File: cpp/memilio/io/result_io.h [hunk 3]
```diff
@@ -86,11 +93,141 @@ class SimulationResult
...
+ * @param ensemble_results Result of each simulation run.
+ * @param ensemble_params Üarameters used for each simulation run.
+ * @param county_ids Ids of the county nodes.
+ * @param result_dir Top level directory for all results of the parameter study.
+ * @param save_single_runs [Default: true] Defines if single run results are written to the disk.
+ * @param save_single_runs [Default: true] Defines if percentiles are written to the disk.
+ * @return Any io errors that occur during writing of the files.
+ */
+template <class Model>
+IOResult<void> save_results(const std::vector<std::vector<TimeSeries<double>>>& ensemble_results,
```
Comment (mknaranja) [inline cpp/memilio/io/result_io.h:136]: needs test

File: cpp/simulations/2020_npis_sarscov2_wildtype_germany.cpp [hunk 4]
```diff
@@ -859,19 +700,31 @@ int main(int argc, char** argv)
...
     std::string data_dir;
     std::string result_dir;
-    if (argc == 4) {
+    bool save_single_runs = true;
+    if (argc == 5) {
+        mode             = RunMode::Save;
+        data_dir         = argv[1];
+        save_dir         = argv[2];
+        result_dir       = argv[3];
+        save_single_runs = argv[4];
```
[the remark this answers is not in the document]
Comment (HenrZu) [reply cpp/simulations/2020_npis_sarscov2_wildtype_germany.cpp:709]: Would be more consistent with the other simulation file

**Review by HenrZu — CHANGES_REQUESTED** [review]

**Review by HenrZu — APPROVED** [review]
Looks good

**Pull request merged**